docs/enhancements/0001-worker-pods.md

```diff
@@ -0,0 +1,229 @@
+            mounts; │                                   │    │     KMM operator │
+reads config from & │       ┌────────────┐              └────┼──────────────────┘
+   writes output to │       │            │                   │
+                    └───────┤ KMM worker ◄───────────────────┘
+                            │    Pod     │ creates & owns
+                            │            │
+                            └──────┬─────┘
+                                   │
+                                   │
+                                   │               ┌──────────────────────┐
+                                   │               │                      │
+                                   └───────────────► registry.example.org │
+                                      pulls from   │                      │
+                                                   └──────────────────────┘
+```
+
+#### The operator
+
+The operator runs two main controllers: the `Module` controller and the `NodeModulesState` controller.
+
+The `Module` controller watches both `Module` and `Node` resources to determine which kmods should be loaded on which
+`Node`.
+It creates one `NodeModulesState` resource per `Node` in the cluster that is targeted to load at least one kmod.  
+`NodeModulesState` resources represent the desired state of all kmods that a certain node should have loaded; they
+contain all the required data to pull kmod images and load the modules properly.  
+The `Module` controller watches updates to the `status` of `NodeModulesState` objects and aggregates that data into the
+`Module` `status`.
+
+The `NodeModulesState` controller watches `NodeModulesState` resources and determines when a worker Pod needs to be
+started on a node by looking at the difference between the desired state (spec) and the current one (status).  
+It prepares a worker configuration file in a `ConfigMap` based on the actions that need to be run on the node and
+creates a Pod mounting that `ConfigMap`.  
+When the Pod has completed, the controller reads the worker's output from the same `ConfigMap` and updates the
+`NodeModulesState` status accordingly.
+
+When updating the `NodeModulesState` status, the controller includes a timestamp corresponding to the last time a worker
+Pod ran on the node.
+This timestamp can be used to verify if a new worker Pod needs to run when the KMM operator restarts; if it older than
+the last transition of the `Node`'s `Ready` condition, then the node may have rebooted and we potentially need to load
+the kernel modules.
+
+#### The worker Pods
+
+Worker Pods are created by the `NodeModulesState` controller.
+They run KMM code that reads the configuration file, pulls and extracts kmod OCI images, uses `modprobe` to load kmods,
+and write some output to a file.
+Both the configuration and the output file are mounted from a `ConfigMap`.
+The `NodeModulesState` creates the `ConfigMaps` as well as the worker Pods.
```

Comment:
Do we consider `NodeModuleState` and the `ConfigMap` as user API or internal implementation?

Reply:
I see that you mentioned in the end of the docs that this is a private API and should not be used by users.

Reply:
customer should only be aware of the Module